Floofbot/Services/RaidProtectionService.cs

```diff
@@ -130,13 +137,63 @@ private void ensureGuildInDictionaries(ulong guildId)
                 userMessageCount.Add(guildId, new Dictionary<ulong, int>());
             if (!punishedUsers.ContainsKey(guildId))
                 punishedUsers.Add(guildId, new List<SocketUser>());
+            if (!lastUserMessageInGuild.ContainsKey(guildId))
+                lastUserMessageInGuild.Add(guildId, new Dictionary<ulong, SocketMessage>());
+        }
+    private bool CheckMessageForFilteredWords(SocketMessage msg, ulong guildId)
+    {
+        bool hasBadWord = _wordFilterService.hasFilteredWord(new FloofDataContext(), msg.Content, guildId, msg.Channel.Id);
+
+        if (hasBadWord)
+        {
+            // add a bad boye point for the user
+            if (userPunishmentCount[guildId].ContainsKey(msg.Author.Id))
+            {
+                userPunishmentCount[guildId][msg.Author.Id] += 1;
+            }
+            else // they were a good boye but now they are not
+            {
+                userPunishmentCount[guildId].Add(msg.Author.Id, 1);
+            }
+            // we run an async task to remove their point after the specified duration
+            UserPunishmentTimeout(guildId, msg.Author.Id);
+
+            SendMessageAndDelete($"{msg.Author.Mention} There was a filtered word in that message. Please be mindful of your language!", msg.Channel);
+
+            Log.Information("User ID " + msg.Author.Id + " triggered the word filter.");
+
+            // we return here because we only need to check for at least one match, doesnt matter if there are more
+            return true;
+        }
+        else
+        {
+            return false;
+        }
+
         }
     private bool CheckUserMessageCount(SocketMessage msg, ulong guildId)
         {
             if (userMessageCount[guildId].ContainsKey(msg.Author.Id))
             {
-                userMessageCount[guildId][msg.Author.Id] += 1;
-                if (userMessageCount[guildId][msg.Author.Id] >= 2) // no more than 2 messages in time frame
+                // update last user message in server
+                if (lastUserMessageInGuild[guildId].ContainsKey(msg.Author.Id))
+                {
+                    lastUserMessageInGuild[guildId][msg.Author.Id] = msg;
+                }
+                // record last user message in server
+                else
+                {
+                    lastUserMessageInGuild[guildId].Add(msg.Author.Id, msg);
+                }
+
+                // compare timestamps of messages
+                TimeSpan timeBetweenMessages = msg.Timestamp - lastUserMessageInGuild[guildId][msg.Author.Id].Timestamp;
```

Comment:
If the `lastUserMessageInGuild` dictionary doesn't contain a user's last message (i.e. this is the first message the bot receives), wouldn't this always be 0 and trigger the following `if` statement?

Reply:
Added in a return statement after the message is added into the dictionary. If this is the first message, we might as well return because they clearly aren't spamming if its the first one